File: app/src/main/java/com/ducktapedapps/updoot/api/AuthAPI.java
package com.ducktapedapps.updoot.api;

import com.ducktapedapps.updoot.model.Token;

import io.reactivex.Single;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.Header;
import retrofit2.http.POST;
import retrofit2.http.Url;

public interface AuthAPI {
    @POST()
    @FormUrlEncoded
    Single<Token> getUserLessToken(
            @Url String auth_url,
            @Header("Authorization") String credentials,
            @Field("grant_type") String grant_type,
            @Field("device_id") String device_id
    );

    @FormUrlEncoded
    @POST
    Single<Token> getUserToken(
            @Url String auth_url,
            @Header("Authorization") String credentials,
            @Field("grant_type") String grant_type,
            @Field("code") String code,
            @Field("redirect_uri") String redirect_uri
    );

    @FormUrlEncoded
    @POST()
    Single<Token> getRefreshedToken(
            @Url String auth_url,
            @Header("Authorization") String credentials,
            @Field("grant_type") String grant_type,
            @Field("refresh_token") String refresh_token
    );
}

File: app/src/main/java/com/ducktapedapps/updoot/api/AuthAPI.kt
package com.ducktapedapps.updoot.api

import com.ducktapedapps.updoot.model.Token
import com.ducktapedapps.updoot.utils.Constants
import io.reactivex.Single
import okhttp3.Credentials
import retrofit2.http.*

interface AuthAPI {
    @POST
    @FormUrlEncoded
    fun getUserLessToken(
            @Url auth_url: String = Constants.TOKEN_ACCESS_URL,
            @Header("Authorization") credentials: String = Credentials.basic(Constants.client_id, ""),
            @Field("grant_type") grant_type: String = Constants.userLess_grantType,
            @Field("device_id") device_id: String
    ): Single<Token>

    @FormUrlEncoded
    @POST
    fun getUserToken(
            @Url auth_url: String = Constants.TOKEN_ACCESS_URL,
            @Header("Authorization") credentials: String = Credentials.basic(Constants.client_id, ""),
            @Field("grant_type") grant_type: String = Constants.user_grantType,
            @Field("code") code: String,
            @Field("redirect_uri") redirect_uri: String = Constants.redirect_uri
    ): Single<Token>

    @FormUrlEncoded
    @POST
    fun getRefreshedToken(
            @Url auth_url: String = Constants.TOKEN_ACCESS_URL,
            @Header("Authorization") credentials: String = Credentials.basic(Constants.client_id, ""),
            @Field("grant_type") grant_type: String = Constants.user_refresh_grantType,
            @Field("refresh_token") refresh_token: String
    ): Single<Token>

}